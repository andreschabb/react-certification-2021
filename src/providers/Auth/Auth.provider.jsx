import React, { useState, useEffect, useContext, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const mockedUser = {
  id: '123',
  name: 'Wizeline',
  avatarUrl:
    'https://media.glassdoor.com/sqll/868055/wizeline-squarelogo-1473976610815.png',
};

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [user, setUser] = useState({});

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setUser(lastAuthState);
    setAuthenticated(isAuthenticated);
  }, []);

  const login = useCallback(async (username, password) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (username === 'wizeline' && password === 'Rocks!') {
          setAuthenticated(true);
          setUser(mockedUser);
          storage.set(AUTH_STORAGE_KEY, mockedUser);
          return resolve(mockedUser);
        }
        return reject(new Error('Username or password invalid'));
      }, 500);
    });
  }, []);

  const logout = useCallback(() => {
    setUser(null);
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, null);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, user }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth };
export default AuthProvider;
