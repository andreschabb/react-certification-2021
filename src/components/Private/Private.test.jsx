import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';

import Private from './Private.component';
import { useAuth } from '../../providers/Auth';

jest.mock('../../providers/Auth');

describe('Private', () => {
  it('should renders with an active sesion', () => {
    useAuth.mockImplementation(() => ({ authenticated: true }));
    const { container } = render(
      <MemoryRouter initialEntries={['/Private']}>
        <Private>
          <div>Private</div>
        </Private>
      </MemoryRouter>
    );

    expect(screen.getByText('Private')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should not renders witout an active sesion', () => {
    useAuth.mockImplementation(() => ({ authenticated: false }));
    render(
      <MemoryRouter initialEntries={['/Private']}>
        <Private>
          <div>Private</div>
        </Private>
      </MemoryRouter>
    );

    expect(screen.queryByText('Private')).toBeFalsy();
  });
});
