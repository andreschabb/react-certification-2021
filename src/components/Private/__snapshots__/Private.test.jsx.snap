// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private should renders with an active sesion 1`] = `
<div>
  <div>
    Private
  </div>
</div>
`;
